### PR TITLE
Fix using a shot_list with multiple shots for sampling

### PR DESCRIPTION
**Pull request by antalszava**
**Context:**

The introduction of partitioned shots allows multiple jobs to be expressed with nearly the same syntax. This allows defining the `shots` attribute of a device as a list of integers.

The circuit runs are separated for each component in the shot vector. The results of these circuit runs are then merged into a single array.

When using `qml.sample`, there can be cases when the shot vector has non-identical elements:
```python
import pennylane as qml

shots_list = [5, 10, 1000]
dev = qml.device("default.qubit", wires=2, shots=shots_list)

@qml.qnode(dev)
def circuit(a):
    qml.RX(a, wires=0)
    return qml.sample(qml.PauliZ(wires=0))

circuit(0.8)
```
Eventually, as per the original logic of using a shot vector, we would like to "merge" (more so use `np.stack`) the result of each circuit run. This is, however, erroneous on the above example:
```python
~/xanadu/pennylane/pennylane/math/numpy_box.py in stack(values, axis)
    104     @wrap_output
    105     def stack(values, axis=0):
--> 106         return np.stack(NumpyBox.unbox_list(values), axis=axis)
    107 
    108     @staticmethod

<__array_function__ internals> in stack(*args, **kwargs)

~/anaconda3/lib/python3.8/site-packages/numpy/core/shape_base.py in stack(arrays, axis, out)
    425     shapes = {arr.shape for arr in arrays}
    426     if len(shapes) != 1:
--> 427         raise ValueError('all input arrays must have the same shape')
    428 
    429     result_ndim = arrays[0].ndim + 1

ValueError: all input arrays must have the same shape
```

**Description of the Change:**
Adds conditions such that the results of sampling with a shot vector are not transformed into an array.

Using `qml.sample` with a shot vector returns a tuple consisting of the results obtained for each component of the shot vector (similar to when having a `return` statement with multiple components).

**Benefits:**
Using `qml.sample` once or multiple times with a shot vector doesn't raise an error.

**Possible Drawbacks:**
N/A

**Related GitHub Issues:**
N/A

### CODE REVIEW

**Comment by josh146** [comment]
Thanks for catching this @antalszava!

Quick question, could this be fixed in less lines by adding the following `if` statement?

```python
# compute the required statistics
if not self.analytic and self._shot_vector is not None:

    results = []
    s1 = 0

    for shot_tuple in self._shot_vector:
        s2 = s1 + np.prod(shot_tuple)
        r = self.statistics(
            circuit.observables, shot_range=[s1, s2], bin_size=shot_tuple.shots
        )
        r = qml.math.squeeze(r)

        if shot_tuple.copies > 1:
            results.extend(r.T)
        else:
            results.append(r)

        s1 = s2

    if not circuit.is_sampled:  # add this if statement here
        results = qml.math.stack(results)
```

With this `if` statement, the stacking is avoided if the circuit is being sampled, and the results are returned as a (ragged) list.

**Comment by antalszava** [comment]
@josh146, thanks for the suggestion! There were two reasons why it seemed less preferable:
* Creating the output emits the `VisibleDeprecationWarning` as we saw in [Strawberry Fields recently when creating ragged numpy arrays](https://github.com/XanaduAI/strawberryfields/pull/564) (we have to specify `'dtype=object'` explicitly to silence the warning)
* Personally, it seems more intuitive to me for a user to receive a tuple of the results, rather than a ragged array (whose `dtype` is also not describing its true content). However, if there are any advantages to having a ragged array, then happy to go with that approach.

**Comment by josh146** [comment]
> Personally, it seems more intuitive to me for a user to receive a tuple of the results, rather than a ragged array (whose dtype is also not describing its true content).

Yes, I agree a tuple is more intuitive - I didn't realize my suggestion _wasn't_ returning a tuple 😆 In that case, I am happy with your solution @antalszava. Although, I might make `has_partitioned_shots` private, not public?

File: pennylane/_qubit_device.py [hunk 1]
```diff
@@ -203,16 +205,18 @@ def execute(self, circuit, **kwargs):
                 if shot_tuple.copies > 1:
                     results.extend(r.T)
                 else:
-                    results.append(r)
+                    results.append(r.T)
```
Comment (antalszava) [inline pennylane/_qubit_device.py:208]: When having multiple `qml.sample` expressions in the return statement, this was required such that the resulting arrays  have the same shape for both when `copies=1` and `copies>1`. (Without this, there's a mismatch, e.g., `shots=[10]` produced `(2,10)` and `shots=[10,10,10]` produced `(10,2)` 3 times).

**Review by josh146 — CHANGES_REQUESTED** [review]
Thanks Antal! Still have a few questions about the implementation, in particularly whether it will break demos/plugins.

File: pennylane/qnode.py [hunk 2]
```diff
@@ -573,7 +573,9 @@ def __call__(self, *args, **kwargs):
             if hasattr(self._original_device, "_state"):
                 self._original_device._state = self.device._state
 
-        if isinstance(self.qfunc_output, Sequence):
+        if isinstance(self.qfunc_output, Sequence) or (
+            self.qtape.is_sampled and self.device._has_partitioned_shots()
```
[the remark this answers is not in the document]
Comment (antalszava) [reply pennylane/qnode.py:577]: `_has_partitioned_shots` is defined in `Device` atm, so all devices should have it

File: pennylane/_qubit_device.py (hunk 1 above)
Comment (antalszava) [reply pennylane/_qubit_device.py:208]: It had no effect on the existing tests

Comment (antalszava) [reply pennylane/_qubit_device.py:208]: As for the demos, [attempted to run the CI](https://app.circleci.com/pipelines/github/PennyLaneAI/qml/2833/workflows/d7e6404a-68a2-4ffe-8d55-62827c7cff62/jobs/2992), though ran into a requirements conflict due to the mismatch in the branches (we require the branch to be `master`).

File: pennylane/_device.py [hunk 3]
```diff
@@ -296,6 +296,16 @@ def shot_vector(self):
         """
         return self._shot_vector
 
+    def _has_partitioned_shots(self):
```
[the remark this answers is not in the document]
Comment (antalszava) [reply pennylane/_device.py:299]: The thing is, that `Device.execute` differs from `QubitDevice.execute`, so defining that behaviour during the device execution would be more demanding.

Would we like to cache to decrease runtime? Calling `_has_partitioned_shots` seems to take around the same time as returning a private attribute through a property:
Screenshot from 2021-04-19 19-06-32

[the remark this answers is not in the document]
Comment (josh146) [reply pennylane/_device.py:299]: Ah, the benchmarking allays my worries 🙂

File: pennylane/_qubit_device.py (hunk 1 above)
Comment (josh146) [reply pennylane/_qubit_device.py:208]: Thanks @antalszava!

File: pennylane/qnode.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (josh146) [reply pennylane/qnode.py:577]: Oh yes I missed this, thanks!

**Review by josh146 — APPROVED** [review]
Thanks Antal for digging into both timing and demos, this allays my fears 🙂 Good to merge this bugfix in prior to the release!

**Pull request merged**